File: 278-first-bad-version/278-first-bad-version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int l = 1, r = n, m;
        while (l <= r) {
            m = l + (r - l) / 2;
            bool bad = isBadVersion(m);
            if (bad && m == 1) return 1;
            bool badBehind = isBadVersion(m - 1);
            if (bad && badBehind) r = m - 1;
            else if (bad && !badBehind) return m;
            else l = m + 1;
        }
        return -1;
    }
};

File: 278-first-bad-version/278-first-bad-version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        l, r = 1, n
        while l <= r:
            m = l + (r - l) // 2
            if m == 1 and isBadVersion(m):
                return 1
            elif isBadVersion(m) and not isBadVersion(m - 1):
                return m
            elif isBadVersion(m):
                r = m - 1
            else:
                l = m + 1
        return -1
        